### DISCUSSION TITLE
Proposal: Allow optional parameters in any location

### DISCUSSION THREAD

**Discussion by ygc369** [dotnet/csharplang#3293] [Language Ideas]
Now the optional parameters in C# function have a limitation: they must be after all mandatory parameters. This limitation is reasonable before C# 7.0. But C# 7.0 introduced "discards" feature (https://github.com/dotnet/csharplang/issues/422), I think that it is possible to remove that limitation by using "discard" feature.
For example:
```C#
static int foo(int a=10, int b) { ...... } // allow optional parameter with default value to locate in front of mandatory parameter
static void main() 
{
    int c = foo(_, 20); // when an optional parameter in front of mandatory parameter is omitted, discard syntax ('_') must be used to occupy its location.
    ......
}
```

**Comment by jnm2** [comment]
Discussion on optional parameters that follow `params`: https://github.com/dotnet/csharplang/issues/1458

**Comment by ygc369** [comment]
@jnm2 
#1458 seems not same with this proposal.

**Comment by jnm2** [comment]
@ygc369 Right; it is related.

**Comment by tannergooding** [comment]
You can already do this today, its just a bit more verbose and not integrated into the language:

```csharp
using System.Runtime.InteropServices;

public class C {
    public void M([Optional, DefaultParameterValue("banana")] string optional, string notOptional) { }
    
    public void N()
    {
        M(notOptional: "a");
    }
}
```
https://sharplab.io/#v2:C4LghgzgtgPgAgBgARwIwDoBKBXAdsASygFN0BJfYgJwHsAHAZWoDcCBjYiAbgFgAofnADMKAExIAwkgDe/JPJQi4AFiQBZABQBtAPJ1CNXGAA2AGiQARYgDMw2Y8AAKYKmBLBqANRPZiGgEQARmBGRv4AlAC6KKjI9AZGZjHIuDTAegkm4TJIAL5yCgXywiiqAHIa4UUy1QrqGqnp+gSGJiBI/mARvHx1+Xy5QA

![image](https://user-images.githubusercontent.com/10487869/77931620-3ed8a100-7261-11ea-93c0-0834c0a7c84a.png)

**Reply by dmchurch** [reply]
Thank you so much for providing this workaround, @tannergooding! I've used this to implement Amit Patel's [cubic hex coordinates](https://www.redblobgames.com/grids/hexagons/#coordinates-cube), which are a three-coordinate system with the constraint that the coordinates must sum to 0; in other words, any two coordinates are enough to define the point:

```cs
using System.Runtime.InteropServices;
public readonly struct HexCoordinate {
    public readonly int q;
    public readonly int r;
    public readonly int s => -q-r;

    public HexCoordinate(int q, int r) => (this.q, this.r) = (q, r);
    public HexCoordinate([Optional] int? q, int r, int s) => (this.q, this.r) = (-r-s, r);
    public HexCoordinate(int q, [Optional] int? r, int s) => (this.q, this.r) = (q, -q-s);

    public static readonly HexCoordinate UnitQR = new(q: 1, r: 1);
    public static readonly HexCoordinate UnitRS = new(r: 1, s: 1);
    public static readonly HexCoordinate UnitSQ = new(s: 1, q: 1);
}
```

**Comment by tannergooding** [comment]
I think the above workaround should be sufficient for the extremely rare scenario (such as certain Interop code) where you want to have an optional parameter that is not last or immediately preceding another optional parameter.

**Comment by ygc369** [comment]
@tannergooding
Thanks for providing the workaround. But I think that my proposal provides a more friendly syntax, and it should not be hard to implement.
